File: cpp/TlvBox.cpp
/*
 *  COPYRIGHT NOTICE  
 *  Copyright (C) 2015, Jhuster, All Rights Reserved
 *  Author: Jhuster(lujun.hust@gmail.com)
 *  
 *  https://github.com/Jhuster/TLV
 *   
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; version 2 of the License.  
 */

#include <string.h>
#include <arpa/inet.h>
#include "TlvBox.h"

namespace tlv
{

TlvBox::TlvBox() : mSerializedBuffer(NULL),mSerializedBytes(0)
{

}

TlvBox::~TlvBox()
{
    if(mSerializedBuffer != NULL) {
        delete[] mSerializedBuffer;
        mSerializedBuffer = NULL;
    }
    mTlvMap.clear();
}

bool TlvBox::Serialize()
{
    if(mSerializedBuffer != NULL) {
        return false;
    }

    int offset = 0;
    mSerializedBuffer = new unsigned char[mSerializedBytes];

    std::map<int,Tlv>::iterator itor;
    for(itor=mTlvMap.begin();itor!=mTlvMap.end();itor++) {
        int type = htonl(itor->second.type);
        memcpy(mSerializedBuffer+offset,&type,sizeof(int));
        offset += sizeof(int);
        int length = htonl(itor->second.length);
        memcpy(mSerializedBuffer+offset,&length,sizeof(int));
        offset += sizeof(int);
        memcpy(mSerializedBuffer+offset,itor->second.value,itor->second.length);
        itor->second.value = mSerializedBuffer+offset;
        offset += itor->second.length;
    }

    return true;
}

bool TlvBox::Parse(const unsigned char *buffer,int buffersize)
{
    if(mSerializedBuffer != NULL) {
        return false;
    }

    unsigned char *cached = new unsigned char[buffersize];
    memcpy(cached,buffer,buffersize);

    int offset = 0;
    while(offset < buffersize) {
        int type = ntohl((*(int *)(cached+offset)));
        offset += sizeof(int);
        int length = ntohl((*(int *)(cached+offset)));
        offset += sizeof(int);
        PutValue(type,cached+offset,length);
        offset += length;
    }

    mSerializedBuffer = cached;
    mSerializedBytes  = buffersize;

    return true;
}

unsigned char * TlvBox::GetSerializedBuffer() const
{
    return mSerializedBuffer;
}

int TlvBox::GetSerializedBytes() const
{
    return mSerializedBytes;
}

bool TlvBox::PutValue(int type,const void *value,int length)
{
    if(mSerializedBuffer != NULL || value == NULL) {
        return false;
    }
    Tlv obj;
    obj.type  = type;
    obj.value = value;
    obj.length = length;
    mTlvMap.insert(std::pair<int,Tlv>(type,obj));
    mSerializedBytes += (sizeof(int)*2+length);
    return true;
}

bool TlvBox::PutCharValue(int type,const char &value)
{
    return PutValue(type,&value,sizeof(char));
}

bool TlvBox::PutShortValue(int type,const short &value)
{
    return PutValue(type,&value,sizeof(short));
}

bool TlvBox::PutIntValue(int type,const int &value)
{
    return PutValue(type,&value,sizeof(int));
}

bool TlvBox::PutLongValue(int type,const long &value)
{
    return PutValue(type,&value,sizeof(long));
}

bool TlvBox::PutLongLongValue(int type,const long long &value)
{
    return PutValue(type,&value,sizeof(long long));
}

bool TlvBox::PutFloatValue(int type,const float &value)
{
    return PutValue(type,&value,sizeof(float));
}

bool TlvBox::PutDoubleValue(int type,const double &value)
{
    return PutValue(type,&value,sizeof(float));
}

bool TlvBox::PutStringValue(int type,const char *value)
{
    return PutValue(type,value,strlen(value)+1);
}

bool TlvBox::PutStringValue(int type,const std::string &value)
{
    int length = value.size()+1;
    return PutValue(type,value.c_str(),length);
}

bool TlvBox::PutBytesValue(int type,const unsigned char *value,int length)
{
    return PutValue(type,value,length);
}

bool TlvBox::PutObjectValue(int type,const TlvBox& value)
{
    return PutValue(type,value.GetSerializedBuffer(),value.GetSerializedBytes());   
}

bool TlvBox::GetValue(int type,const void **value,int& length) const
{
    std::map<int,Tlv>::const_iterator itor = mTlvMap.find(type);
    if(itor == mTlvMap.end()) {
        return false;
    }
    *value = itor->second.value;
    length = itor->second.length;
    return true;
}

bool TlvBox::GetCharValue(int type,char &value) const
{
    Tlv obj;
    if(!GetValue(type,&obj.value,obj.length)) {
        return false;
    }
    value = (*(char *)(obj.value));
    return true;
}

bool TlvBox::GetShortValue(int type,short &value) const
{
    Tlv obj;
    if(!GetValue(type,&obj.value,obj.length)) {
        return false;
    }
    value = (*(short *)(obj.value));
    return true;
}

bool TlvBox::GetIntValue(int type,int &value) const
{
    Tlv obj;
    if(!GetValue(type,&obj.value,obj.length)) {
        return false;
    }
    value = (*(int *)(obj.value));
    return true;
}

bool TlvBox::GetLongValue(int type,long &value) const
{
    Tlv obj;
    if(!GetValue(type,&obj.value,obj.length)) {
        return false;
    }
    value = (*(long *)(obj.value));
    return true;
}

bool TlvBox::GetLongLongValue(int type,long long &value) const
{
    Tlv obj;
    if(!GetValue(type,&obj.value,obj.length)) {
        return false;
    }
    value = (*(long long *)(obj.value));
    return true;
}

bool TlvBox::GetFloatValue(int type,float& value) const
{
    Tlv obj;
    if(!GetValue(type,&obj.value,obj.length)) {
        return false;
    }
    value = (*(float *)(obj.value));
    return true;
}

bool TlvBox::GetDoubleValue(int type,double& value) const
{
    Tlv obj;
    if(!GetValue(type,&obj.value,obj.length)) {
        return false;
    }
    value = (*(double *)(obj.value));
    return true;
}

bool TlvBox::GetStringValue(int type,char *value,int &length) const
{
    return GetBytesValue(type,(unsigned char *)value,length);
}

bool TlvBox::GetStringValue(int type,std::string &value) const
{
    Tlv obj;
    if(!GetValue(type,&obj.value,obj.length)) {
        return false;
    }
    value = (char *)obj.value;
    return true;
}

bool TlvBox::GetBytesValue(int type,unsigned char *value,int &length) const
{
    Tlv obj;
    if(!GetValue(type,&obj.value,obj.length) || length < obj.length) {
        return false;
    }
    memset(value,0,length);
    length = obj.length;
    memcpy(value,obj.value,obj.length);
    return true;
}

bool TlvBox::GetBytesValuePtr(int type,unsigned char **value,int &length) const
{
    Tlv obj;
    if(!GetValue(type,&obj.value,obj.length)) {
        return false;
    }
    *value = (unsigned char *)obj.value;
    length = obj.length;
    return true;
}

bool TlvBox::GetObjectValue(int type,TlvBox& value) const
{
    Tlv obj;
    if(!GetValue(type,&obj.value,obj.length)) {
        return false;
    }
    if(!value.Parse((const unsigned char *)obj.value,obj.length)) {
        return false;   
    }    
    return true;
}


} //namespace 

File: Android/TlvBox.java
/*
 *  COPYRIGHT NOTICE  
 *  Copyright (C) 2015, Jhuster, All Rights Reserved
 *  Author: Jhuster(lujun.hust@gmail.com)
 *  
 *  https://github.com/Jhuster/TLV
 *   
 *  This library is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published
 *  by the Free Software Foundation; either version 2.1 of the License, 
 *  or (at your option) any later version.
 */
package com.jhuster.testcode;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;

import android.util.SparseArray;

public class TlvBox {
    
    private static final ByteOrder DEFAULT_BYTE_ORDER = ByteOrder.LITTLE_ENDIAN;
    
    private SparseArray<byte[]> mObjects;
    private int mTotalBytes = 0;

    public TlvBox() {
        mObjects = new SparseArray<byte[]>();
    }
    
    public static TlvBox parse(byte[] buffer, int offset, int length) {
        
        TlvBox box = new TlvBox();
        
        int parsed = 0;
        while (parsed < length) {
            int type = ByteBuffer.wrap(buffer, offset + parsed, 4).order(DEFAULT_BYTE_ORDER).getInt();
            parsed += 4;
            int size = ByteBuffer.wrap(buffer, offset+  parsed, 4).order(DEFAULT_BYTE_ORDER).getInt();
            parsed += 4;
            byte[] value = new byte[size];
            System.arraycopy(buffer, offset + parsed, value, 0, size);
            box.putBytesValue(type, value);
            parsed += size;
        }        
        
        return box;
    }
    
    public byte[] serialize() {
        int offset = 0;
        byte[] result = new byte[mTotalBytes];  
        for(int i=0; i<mObjects.size(); i++) {
            int key = mObjects.keyAt(i);
            byte[] bytes = mObjects.get(key);
            byte[] type   = ByteBuffer.allocate(4).order(DEFAULT_BYTE_ORDER).putInt(key).array();
            byte[] length = ByteBuffer.allocate(4).order(DEFAULT_BYTE_ORDER).putInt(bytes.length).array();
            System.arraycopy(type, 0, result, offset, type.length);
            offset += 4;
            System.arraycopy(length, 0, result, offset, length.length);
            offset += 4;
            System.arraycopy(bytes, 0, result, offset, bytes.length);
            offset += bytes.length;
        }
        return result;
    }
    
    public void putByteValue(int type,byte value) {
        byte[] bytes = new byte[1];        
        bytes[0] = value;
        putBytesValue(type, bytes);
    }
        
    public void putShortValue(int type,short value) {
        byte[] bytes = ByteBuffer.allocate(2).order(DEFAULT_BYTE_ORDER).putShort(value).array();
        putBytesValue(type, bytes);
    }
    
    public void putIntValue(int type,int value) {
        byte[] bytes = ByteBuffer.allocate(4).order(DEFAULT_BYTE_ORDER).putInt(value).array();
        putBytesValue(type, bytes);
    }
    
    public void putLongValue(int type,long value) {
        byte[] bytes = ByteBuffer.allocate(8).order(DEFAULT_BYTE_ORDER).putLong(value).array();
        putBytesValue(type, bytes);
    }
    
    public void putFloatValue(int type,float value) {
        byte[] bytes = ByteBuffer.allocate(4).order(DEFAULT_BYTE_ORDER).putFloat(value).array();
        putBytesValue(type, bytes);
    }
    
    public void putDoubleValue(int type,double value) {
        byte[] bytes = ByteBuffer.allocate(8).order(DEFAULT_BYTE_ORDER).putDouble(value).array();
        putBytesValue(type, bytes);
    }
    
    public void putStringValue(int type,String value) {         
        putBytesValue(type,value.getBytes());        
    }

    public void putObjectValue(int type,TlvBox value) {        
        putBytesValue(type,value.serialize());
    }
    
    public void putBytesValue(int type,byte[] value) {
        mObjects.put(type, value);
        mTotalBytes += value.length + 8;
    }
    
    public Byte getByteValue(int type) {
        byte[] bytes = mObjects.get(type);
        if (bytes == null) {
            return null;
        }
        return bytes[0];
    }
        
    public Short getShortValue(int type) {
        byte[] bytes = mObjects.get(type);
        if (bytes == null) {
            return null;
        }
        return ByteBuffer.wrap(bytes).order(DEFAULT_BYTE_ORDER).getShort();
    }
    
    public Integer getIntValue(int type) {
        byte[] bytes = mObjects.get(type);
        if (bytes == null) {
            return null;
        }
        return ByteBuffer.wrap(bytes).order(DEFAULT_BYTE_ORDER).getInt();
    }
    
    public Long getLongValue(int type) {
        byte[] bytes = mObjects.get(type);
        if (bytes == null) {
            return null;
        }
        return ByteBuffer.wrap(bytes).order(DEFAULT_BYTE_ORDER).getLong();
    }
    
    public Float getFloatValue(int type) {
        byte[] bytes = mObjects.get(type);
        if (bytes == null) {
            return null;
        }
        return ByteBuffer.wrap(bytes).order(DEFAULT_BYTE_ORDER).getFloat();
    }
    
    public Double getDoubleValue(int type) {
        byte[] bytes = mObjects.get(type);
        if (bytes == null) {
            return null;
        }
        return ByteBuffer.wrap(bytes).order(DEFAULT_BYTE_ORDER).getDouble();
    }
    
    public TlvBox getObjectValue(int type) {
        byte[] bytes = mObjects.get(type);
        if (bytes == null) {
            return null;
        }
        return TlvBox.parse(bytes, 0, bytes.length);
    }
    
    public String getStringValue(int type) {
        byte[] bytes = mObjects.get(type);
        if (bytes == null) {
            return null;
        }
        return new String(bytes).trim();
    }
    
    public byte[] getBytesValue(int type) {
        byte[] bytes = mObjects.get(type);
        return bytes;
    }
}
